Backport of:

From 76cc91d65be3b698e92190c153602dc32c5b07c7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bastien=20ROUCARI=C3=88S?= <roucaries.bastien@gmail.com>
Date: Wed, 24 Aug 2016 15:33:15 +0200
Subject: [PATCH] Avoid DOS by using correct locking

There is a bug in the locking implentation (RAII was the intended C++ idiom) that has been fixed upstream.

http://git.imagemagick.org/repos/ImageMagick/commit/5cbe21ed2728da0e611154d2f8e41bb63095a62c

Unfortunately, the commit message is empty...

In the unfixed code, the mutex acquisition has no effect and doesn't prevent concurrent access to ref counters.

This bug generates a lot of crashes when Magick++ is used with multi-threaded applications

Minimal test case:
$ cat bug.cc
using namespace Magick;
int main(int argc, char **argv)
{
  Image plop("/usr/share/pixmaps/debian-logo.png");
  for (int i = 0 ; i < 10000 ; ++i )
    {
      Image meh(plop);
    }
    return 0;
}
$ g++ -fopenmp $(pkg-config --cflags --libs Magick++) bug.cc -o bug
$ ./bug
bug: ../../magick/image.c:1106: DestroyImageInfo: Assertion `image_info->signature == 0xabacadabUL' failed.
Aborted
$ ./bug
bug: ../../magick/image.c:1106: DestroyImageInfo: Assertion `image_info->signature == 0xabacadabUL' failed.
terminate called after throwing an instance of 'Magick::ErrorOption'
  what():  Magick: mutex lock failed (Invalid argument)
Aborted

crash may vary depending on which race is triggered.

Thanks to Guillaume Gimenez (guillaume@blackmilk.fr)

(cherry picked from commit 5cbe21ed2728da0e611154d2f8e41bb63095a62c)
origin: upstream, https://github.com/ImageMagick/ImageMagick/commit/5cbe21ed2728da0e611154d2f8e41bb63095a62c
bug-debian: https://bugs.debian.org/834163
---
 Magick++/lib/Blob.cpp  | 12 ++++++------
 Magick++/lib/Image.cpp | 14 +++++++-------
 2 files changed, 13 insertions(+), 13 deletions(-)

Index: imagemagick-6.6.9.7/Magick++/lib/Blob.cpp
===================================================================
--- imagemagick-6.6.9.7.orig/Magick++/lib/Blob.cpp	2016-11-15 15:59:03.812735283 -0500
+++ imagemagick-6.6.9.7/Magick++/lib/Blob.cpp	2016-11-15 15:59:03.808735235 -0500
@@ -35,7 +35,7 @@
   : _blobRef(blob_._blobRef)
 {
   // Increase reference count
-  Lock( &_blobRef->_mutexLock );
+ Lock lock(&_blobRef->_mutexLock);
   ++_blobRef->_refCount;
 }
 
@@ -44,7 +44,7 @@
 {
   bool doDelete = false;
   {
-    Lock( &_blobRef->_mutexLock );
+   Lock lock(&_blobRef->_mutexLock);
     if ( --_blobRef->_refCount == 0 )
       doDelete = true;
   }
@@ -63,12 +63,12 @@
   if(this != &blob_)
     {
       {
-        Lock( &blob_._blobRef->_mutexLock );
+       Lock lock(&blob_._blobRef->_mutexLock);
         ++blob_._blobRef->_refCount;
       }
       bool doDelete = false;
       {
-        Lock( &_blobRef->_mutexLock );
+       Lock lock(&_blobRef->_mutexLock);
         if ( --_blobRef->_refCount == 0 )
           doDelete = true;
       }
@@ -118,7 +118,7 @@
 {
   bool doDelete = false;
   {
-    Lock( &_blobRef->_mutexLock );
+   Lock lock( &_blobRef->_mutexLock );
     if ( --_blobRef->_refCount == 0 )
       doDelete = true;
   }
@@ -140,7 +140,7 @@
 {
   bool doDelete = false;
   {
-    Lock( &_blobRef->_mutexLock );
+   Lock lock(&_blobRef->_mutexLock);
     if ( --_blobRef->_refCount == 0 )
       doDelete = true;
   }
Index: imagemagick-6.6.9.7/Magick++/lib/Image.cpp
===================================================================
--- imagemagick-6.6.9.7.orig/Magick++/lib/Image.cpp	2016-11-15 15:59:03.812735283 -0500
+++ imagemagick-6.6.9.7/Magick++/lib/Image.cpp	2016-11-15 15:59:03.808735235 -0500
@@ -286,7 +286,7 @@
 {
   bool doDelete = false;
   {
-    Lock( &_imgRef->_mutexLock );
+   Lock lock(&_imgRef->_mutexLock);
     if ( --_imgRef->_refCount == 0 )
       doDelete = true;
   }
@@ -3535,7 +3535,7 @@
 
 std::string Magick::Image::signature ( const bool force_ ) const
 {
-  Lock( &_imgRef->_mutexLock );
+ Lock lock(&_imgRef->_mutexLock);
 
   // Re-calculate image signature if necessary
   if ( force_ ||
@@ -3932,7 +3932,7 @@
 Magick::Image::Image( const Image & image_ )
   : _imgRef(image_._imgRef)
 {
-  Lock( &_imgRef->_mutexLock );
+   Lock lock(&_imgRef->_mutexLock);
 
   // Increase reference count
   ++_imgRef->_refCount;
@@ -3944,13 +3944,13 @@
   if( this != &image_ )
     {
       {
-        Lock( &image_._imgRef->_mutexLock );
+       Lock lock(&image_._imgRef->_mutexLock);
         ++image_._imgRef->_refCount;
       }
 
       bool doDelete = false;
       {
-        Lock( &_imgRef->_mutexLock );
+       Lock lock(&_imgRef->_mutexLock);
         if ( --_imgRef->_refCount == 0 )
           doDelete = true;
       }
@@ -4173,7 +4173,7 @@
     image = AcquireImage(constImageInfo());
 
   {
-    Lock( &_imgRef->_mutexLock );
+   Lock lock(&_imgRef->_mutexLock);
 
     if ( _imgRef->_refCount == 1 )
       {
@@ -4199,7 +4199,7 @@
 void Magick::Image::modifyImage( void )
 {
   {
-    Lock( &_imgRef->_mutexLock );
+   Lock lock(&_imgRef->_mutexLock);
     if ( _imgRef->_refCount == 1 )
       {
         // De-register image and return
@@ -4234,7 +4234,7 @@
 // Register image with image registry or obtain registration id
 ssize_t Magick::Image::registerId( void )
 {
-  Lock( &_imgRef->_mutexLock );
+   Lock lock(&_imgRef->_mutexLock);
   if( _imgRef->id() < 0 )
     {
       char id[MaxTextExtent];
